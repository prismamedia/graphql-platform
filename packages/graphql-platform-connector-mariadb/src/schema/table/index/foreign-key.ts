import * as core from '@prismamedia/graphql-platform';
import * as utils from '@prismamedia/graphql-platform-utils';
import { Memoize } from '@prismamedia/memoize';
import { escapeIdentifier } from '../../../escaping.js';
import type { MariaDBConnector } from '../../../index.js';
import type { Table } from '../../table.js';
import { AbstractIndex } from '../abstract-index.js';
import { ReferenceColumn } from '../column/reference.js';

export interface ForeignKeyIndexConfig {
  /**
   * Optional, the index's name
   */
  name?: utils.Nillable<string>;
}

/**
 * @see https://mariadb.com/kb/en/foreign-keys/
 */
export class ForeignKeyIndex extends AbstractIndex {
  public readonly config?: ForeignKeyIndexConfig;
  public readonly configPath: utils.Path;

  public constructor(
    table: Table,
    public readonly edge: core.Edge<any, MariaDBConnector>,
  ) {
    super(table);

    // config
    {
      this.config = edge.config.foreignKey;
      this.configPath = utils.addPath(edge.configPath, 'foreignKey');

      utils.assertNillablePlainObjectConfig(this.config, this.configPath);
    }
  }

  @Memoize()
  public get referencedTable(): Table {
    return this.table.schema.getTableByNode(this.edge.head);
  }

  @Memoize()
  public get columns(): ReadonlyArray<ReferenceColumn> {
    return this.table.getColumnTreeByEdge(this.edge).columns;
  }

  @Memoize()
  public get name(): string {
    const nameConfig = this.config?.name;
    const nameConfigPath = utils.addPath(this.configPath, 'name');

    if (nameConfig) {
      if (typeof nameConfig !== 'string') {
        throw new utils.UnexpectedConfigError('a string', nameConfig, {
          path: nameConfigPath,
        });
      }

      if (nameConfig.length > 64) {
        throw new utils.UnexpectedConfigError(
          'an identifier shorter than 64 characters',
          nameConfig,
          { path: nameConfigPath },
        );
      }

      return nameConfig;
    }

    return this.table.schema.namingStrategy.getForeignKeyIndexName(
      this.table.name,
      this.edge,
      this.columns,
    );
  }

  /**
   * @see https://mariadb.com/kb/en/create-table/#foreign-key
   */
  @Memoize()
  public get definition(): string {
    return [
      'FOREIGN KEY',
      escapeIdentifier(this.name),
      `(${this.columns.map(({ name }) => escapeIdentifier(name)).join(',')})`,
      'REFERENCES',
      escapeIdentifier(this.referencedTable.name),
      `(${this.columns
        .map(({ referencedColumn: { name } }) => escapeIdentifier(name))
        .join(',')})`,
      'ON UPDATE RESTRICT',
      'ON DELETE RESTRICT',
    ]
      .filter(Boolean)
      .join(' ');
  }
}
