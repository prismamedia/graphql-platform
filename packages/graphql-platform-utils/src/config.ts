export * from './config/name.js';
export * from './config/optional-deprecation.js';
export * from './config/optional-description.js';
export * from './config/optional-flag.js';
