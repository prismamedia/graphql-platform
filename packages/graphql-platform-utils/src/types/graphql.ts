export {
  FieldConfig as GraphQLFieldConfig,
  FieldConfigMap as GraphQLFieldConfigMap,
  FieldResolver as GraphQLFieldResolver,
  FieldResolverArgs as GraphQLFieldResolverArgs,
} from './graphql/field';
export { OperationType as GraphQLOperationType } from './graphql/operation-type';
