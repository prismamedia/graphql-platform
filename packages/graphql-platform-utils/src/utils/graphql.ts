export * from './graphql/get-enum-type';
export * from './graphql/list-decorator';
export * from './graphql/non-null-decorator';
export * from './graphql/parse-resolver-info';
