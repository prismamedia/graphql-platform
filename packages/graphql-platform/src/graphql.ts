import type * as graphql from 'graphql';
import type { ConnectorInterface } from './connector-interface.js';
import type { GraphQLPlatform } from './index.js';

export type GPBoundGraphQLFieldResolver<
  TRequestContext extends object,
  TConnector extends ConnectorInterface,
  TSource,
  TArgs = any,
  TResult = unknown,
> = (
  this: GraphQLPlatform<TRequestContext, TConnector>,
  source: TSource,
  args: TArgs,
  context: TRequestContext,
  info: graphql.GraphQLResolveInfo,
) => TResult;

export interface GPBoundGraphQLFieldConfig<
  TRequestContext extends object,
  TConnector extends ConnectorInterface,
  TSource,
  TArgs = any,
  TResult = unknown,
> extends graphql.GraphQLFieldConfig<TSource, TRequestContext, TArgs> {
  resolve?: GPBoundGraphQLFieldResolver<
    TRequestContext,
    TConnector,
    TSource,
    TArgs,
    TResult
  >;
}
