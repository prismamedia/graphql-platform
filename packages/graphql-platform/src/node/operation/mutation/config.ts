import type * as utils from '@prismamedia/graphql-platform-utils';
import type { ConnectorInterface } from '../../../connector-interface.js';
import type { CreationConfig } from './abstract-creation.js';
import type { DeletionConfig } from './abstract-deletion.js';
import type { UpdateConfig } from './abstract-update.js';

export type {
  CreationConfig,
  PostCreateArgs,
  PreCreateArgs,
} from './abstract-creation.js';
export type {
  DeletionConfig,
  PostDeleteArgs,
  PreDeleteArgs,
} from './abstract-deletion.js';
export type {
  PostUpdateArgs,
  PreUpdateArgs,
  UpdateConfig,
} from './abstract-update.js';

export interface MutationConfig<
  TRequestContext extends object,
  TConnector extends ConnectorInterface,
  TContainer extends object,
> {
  [utils.MutationType.CREATION]: CreationConfig<
    TRequestContext,
    TConnector,
    TContainer
  >;
  [utils.MutationType.UPDATE]: UpdateConfig<
    TRequestContext,
    TConnector,
    TContainer
  >;
  [utils.MutationType.DELETION]: DeletionConfig<
    TRequestContext,
    TConnector,
    TContainer
  >;
}
