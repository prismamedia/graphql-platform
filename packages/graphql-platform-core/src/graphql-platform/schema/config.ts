import { GraphQLOperationType } from '@prismamedia/graphql-platform-utils';
import { GraphQLFieldConfigMap, GraphQLObjectType, GraphQLSchemaConfig as BaseGraphQLSchemaConfig } from 'graphql';
import { OperationType } from '../operation';
import { ResourceMap } from '../resource/map';

export interface GraphQLSchemaConfigConfig
  extends Omit<BaseGraphQLSchemaConfig, 'mutation' | 'query' | 'subscription'> {
  resourceMap: ResourceMap;
  queryMap: GraphQLFieldConfigMap<any, any>;
  mutationMap: GraphQLFieldConfigMap<any, any>;
}

export class GraphQLSchemaConfig implements BaseGraphQLSchemaConfig {
  public constructor(readonly config: GraphQLSchemaConfigConfig) {}

  public getOperationTypeFieldMap<TType extends OperationType>(type: TType): GraphQLFieldConfigMap<any, any> {
    const fields: GraphQLFieldConfigMap<any, any> = {};

    for (const resource of this.config.resourceMap.values()) {
      for (const operation of resource.getOperations(type)) {
        if (operation.isPublic()) {
          fields[operation.name] = operation.getGraphQLFieldConfig();
        }
      }
    }

    return fields;
  }

  public get mutation() {
    return new GraphQLObjectType({
      name: 'Mutation',
      fields: { ...this.getOperationTypeFieldMap(GraphQLOperationType.Mutation), ...this.config.mutationMap },
    });
  }

  public get query() {
    return new GraphQLObjectType({
      name: 'Query',
      fields: { ...this.getOperationTypeFieldMap(GraphQLOperationType.Query), ...this.config.queryMap },
    });
  }

  public get subscription() {
    return undefined;
  }
}
