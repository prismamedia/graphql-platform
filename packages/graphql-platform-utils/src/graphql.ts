import * as graphql from 'graphql';
import { getOptionalFlag, OptionalFlag } from './config.js';
import { getEnumKeys } from './enum.js';
import { UnexpectedValueError } from './error.js';
import { indefinite } from './indefinite.js';
import { isNil, type Nillable } from './nil.js';
import type { Path } from './path.js';
import { isPlainObject, type PlainObject } from './plain-object.js';

export const isGraphQLResolveInfo = (
  maybeGraphQLResolveInfo: unknown,
): maybeGraphQLResolveInfo is graphql.GraphQLResolveInfo =>
  isPlainObject(maybeGraphQLResolveInfo) &&
  typeof maybeGraphQLResolveInfo['fieldName'] === 'string' &&
  Array.isArray(maybeGraphQLResolveInfo['fieldNodes']) &&
  graphql.isOutputType(maybeGraphQLResolveInfo['returnType']) &&
  graphql.isOutputType(maybeGraphQLResolveInfo['parentType']);

export function assertGraphQLResolveInfo(
  maybeGraphQLResolveInfo: unknown,
  path?: Path,
): asserts maybeGraphQLResolveInfo is graphql.GraphQLResolveInfo {
  if (!isGraphQLResolveInfo(maybeGraphQLResolveInfo)) {
    throw new UnexpectedValueError(
      `a GraphQLResolveInfo`,
      maybeGraphQLResolveInfo,
      { path },
    );
  }
}

export const isGraphQLASTNode = <TKind extends graphql.Kind>(
  maybeGraphQLASTNode: unknown,
  kind: TKind,
): maybeGraphQLASTNode is graphql.ASTKindToNode[TKind] =>
  isPlainObject(maybeGraphQLASTNode) && maybeGraphQLASTNode['kind'] === kind;

export function assertGraphQLASTNode<TKind extends graphql.Kind>(
  maybeGraphQLASTNode: unknown,
  kind: TKind,
  path?: Path,
): asserts maybeGraphQLASTNode is graphql.ASTKindToNode[TKind] {
  if (!isGraphQLASTNode(maybeGraphQLASTNode, kind)) {
    throw new UnexpectedValueError(`a GraphQL ${kind}`, maybeGraphQLASTNode, {
      path,
    });
  }
}

export const parseGraphQLScalarValue = <TInternal>(
  type: graphql.GraphQLScalarType<TInternal, any>,
  maybeScalarValue: unknown,
  path?: Path,
): Nillable<TInternal> => {
  if (isNil(maybeScalarValue)) {
    return maybeScalarValue;
  }

  try {
    return type.parseValue(maybeScalarValue);
  } catch (error) {
    throw new UnexpectedValueError(indefinite(type.name), maybeScalarValue, {
      cause: error,
      path,
    });
  }
};

export const parseGraphQLEnumValue = <TInternal = any>(
  type: graphql.GraphQLEnumType,
  maybeEnumValue: unknown,
  path?: Path,
): Nillable<TInternal> => {
  if (isNil(maybeEnumValue)) {
    return maybeEnumValue;
  }

  const enumValue = type
    .getValues()
    .find(({ value }) => value === maybeEnumValue);

  if (!enumValue) {
    throw new UnexpectedValueError(
      `${indefinite(type.name)} (= a value among "${type
        .getValues()
        .map(({ value }) => value)
        .join(', ')}")`,
      maybeEnumValue,
      { path },
    );
  }

  return enumValue.value;
};

export const parseGraphQLLeafValue = (
  type: graphql.GraphQLLeafType,
  maybeLeafValue: unknown,
  path?: Path,
): any => {
  if (isNil(maybeLeafValue)) {
    return maybeLeafValue;
  }

  return graphql.isScalarType(type)
    ? parseGraphQLScalarValue(type, maybeLeafValue, path)
    : parseGraphQLEnumValue(type, maybeLeafValue, path);
};

export type CreateGraphQLEnumTypeOptions = {
  description?: Nillable<string>;
  useKeyAsValue?: OptionalFlag;
};

export const createGraphQLEnumType = (
  name: string,
  enumerable: PlainObject,
  options?: CreateGraphQLEnumTypeOptions,
): graphql.GraphQLEnumType =>
  new graphql.GraphQLEnumType({
    name,
    description: options?.description,
    values: Object.fromEntries(
      getEnumKeys(enumerable).map<[string, graphql.GraphQLEnumValueConfig]>(
        (key) => [
          key,
          {
            value: getOptionalFlag(options?.useKeyAsValue, false)
              ? undefined
              : enumerable[key],
          },
        ],
      ),
    ),
  });
