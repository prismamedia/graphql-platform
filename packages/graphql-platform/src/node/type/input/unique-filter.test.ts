import { GraphQLInputObjectType, printType } from 'graphql';
import { GraphQLPlatform } from '../../../index.js';
import { MyGP, nodeNames, nodes } from '../../../__tests__/config.js';
import {
  NodeUniqueFilterInputType,
  NodeUniqueFilterInputValue,
} from './unique-filter.js';

describe('NodeUniqueFilterInputType', () => {
  let gp: MyGP;

  beforeAll(() => {
    gp = new GraphQLPlatform({ nodes });
  });

  describe('Definition', () => {
    it.each(nodeNames)('%s has a unique filter input type', (nodeName) => {
      const node = gp.getNodeByName(nodeName);

      const uniqueFilterInputType = node.uniqueFilterInputType;
      expect(uniqueFilterInputType).toBeInstanceOf(NodeUniqueFilterInputType);

      if (node.isPublic()) {
        expect(uniqueFilterInputType.getGraphQLInputType()).toBeInstanceOf(
          GraphQLInputObjectType,
        );
        expect(
          printType(uniqueFilterInputType.getGraphQLInputType()),
        ).toMatchSnapshot(uniqueFilterInputType.name);
      } else {
        expect(() => uniqueFilterInputType.getGraphQLInputType()).toThrowError(
          `The "${nodeName}UniqueFilterInput" input type is private`,
        );
      }
    });
  });

  describe('Parser', () => {
    describe('Fails', () => {
      it.each([
        [
          'Article',
          { id_is_null: true },
          'Expects not to contain the extra key(s): id_is_null',
        ],
        ['Article', { id: null }, '/id - Expects a non-null "UUIDv4"'],
        ['Article', { id: 123 }, `/id - Expects an "UUIDv4", got: 123`],
        [
          'Article',
          { category: { parent: { title: null } }, slug: 'you-re-welcome' },
          '/category/parent - Expects not to contain the extra key(s): title',
        ],
      ])(
        '%sUniqueFilterInput.parseValue(%p) throws the error %p',
        (nodeName, value, error) => {
          const node = gp.getNodeByName(nodeName);
          const uniqueFilterInputType = node.uniqueFilterInputType;

          expect(() => uniqueFilterInputType.parseValue(value)).toThrowError(
            error,
          );
        },
      );
    });

    describe('Works', () => {
      it.each<[string, NodeUniqueFilterInputValue, NodeUniqueFilterInputValue]>(
        [
          ['Article', undefined, undefined],
          ['Article', null, null],
          ['Article', { _id: 123 }, { _id: 123 }],
          [
            'Article',
            { id: 'e22205cc-7d8e-4772-a46d-528a29fb81f2' },
            { id: 'e22205cc-7d8e-4772-a46d-528a29fb81f2' },
          ],
          [
            'Category',
            {
              parent: { id: '6684d029-0016-4615-b1e7-f7f0087dbf11' },
              order: 789,
            },
            {
              parent: { id: '6684d029-0016-4615-b1e7-f7f0087dbf11' },
              order: 789,
            },
          ],
          [
            'Category',
            { parent: null, slug: 'root' },
            { parent: null, slug: 'root' },
          ],
          [
            'Category',
            {
              parent: {
                _id: 999,
                slug: 'root',
                id: 'd038d5cf-815e-4f8d-8099-1def0bdec246',
              },
              order: 789,
            },
            {
              parent: {
                _id: 999,
              },
              order: 789,
            },
          ],
        ],
      )(
        '%sUniqueFilterInput.parseValue(%p) = %p',
        (nodeName, inputValue, parsedValue) => {
          const uniqueFilterInputType =
            gp.getNodeByName(nodeName).uniqueFilterInputType;

          expect(uniqueFilterInputType.parseValue(inputValue)).toEqual(
            parsedValue,
          );
        },
      );
    });
  });
});
