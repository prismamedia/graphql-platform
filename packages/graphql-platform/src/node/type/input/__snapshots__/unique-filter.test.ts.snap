// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NodeUniqueFilterInputType Definition Article has a unique filter input type: ArticleUniqueFilterInput 1`] = `
""""
Identifies exactly one "Article" given one of the following combination of components' value:
- id
- category / slug
"""
input ArticleUniqueFilterInput {
  """This UUID identifies an Article publicly"""
  id: UUIDv4
  category: CategoryUniqueFilterInput
  slug: NonEmptyTrimmedString
}"
`;

exports[`NodeUniqueFilterInputType Definition ArticleTag has a unique filter input type: ArticleTagUniqueFilterInput 1`] = `
""""
Identifies exactly one "ArticleTag" given one of the following combination of components' value:
- article / tag
- article / order
"""
input ArticleTagUniqueFilterInput {
  article: ArticleUniqueFilterInput!
  tag: TagUniqueFilterInput
  order: UnsignedInt
}"
`;

exports[`NodeUniqueFilterInputType Definition ArticleTagModeration has a unique filter input type: ArticleTagModerationUniqueFilterInput 1`] = `
""""
Identifies exactly one "ArticleTagModeration" given one of the following combination of components' value:
- articleTag / moderator
"""
input ArticleTagModerationUniqueFilterInput {
  articleTag: ArticleTagUniqueFilterInput!
  moderator: UserUniqueFilterInput!
}"
`;

exports[`NodeUniqueFilterInputType Definition Category has a unique filter input type: CategoryUniqueFilterInput 1`] = `
""""
Identifies exactly one "Category" given one of the following combination of components' value:
- id
- parent / slug
- parent / order
"""
input CategoryUniqueFilterInput {
  id: UUIDv4
  parent: CategoryUniqueFilterInput
  slug: NonEmptyTrimmedString
  order: UnsignedInt
}"
`;

exports[`NodeUniqueFilterInputType Definition Tag has a unique filter input type: TagUniqueFilterInput 1`] = `
""""
Identifies exactly one "Tag" given one of the following combination of components' value:
- id
- slug
"""
input TagUniqueFilterInput {
  id: UUIDv4
  slug: NonEmptyTrimmedString
}"
`;

exports[`NodeUniqueFilterInputType Definition User has a unique filter input type: UserUniqueFilterInput 1`] = `
""""
Identifies exactly one "User" given one of the following combination of components' value:
- id
- username
"""
input UserUniqueFilterInput {
  id: UUIDv4
  username: NonEmptyTrimmedString
}"
`;

exports[`NodeUniqueFilterInputType Definition UserProfile has a unique filter input type: UserProfileUniqueFilterInput 1`] = `
""""
Identifies exactly one "UserProfile" given one of the following combination of components' value:
- user
"""
input UserProfileUniqueFilterInput {
  user: UserUniqueFilterInput!
}"
`;
